### DISCUSSION TITLE
Not able to use async/await operations in browser mode

### DISCUSSION THREAD

**Discussion by rincew1nd** [AvaloniaUI/Avalonia#16410] [Q&A] [answered]
Hello.

I'm using latest cross platform template (Desktop and Browser builds) without ReactiveUI (with community toolkit).
At some point i decided to add a browser support for the desktop application and found out that async/await operations are not working in browser builds.
It's working flawlessly in desktop mode, but in browser mode calls are lost after awaiting them and nothing happens in network activity nor debug.
What am I doing wrong there? Is there any difference in workflow with async/await in browser build?

Here is samples of code that I'm using:
- MainView.axaml
``` 
  <Design.DataContext>
    <vm:MainViewModel />
  </Design.DataContext>
  <StackPanel>
      <Button Content="ViewCommand" Command="{Binding CallAsync}" />
      <Button Content="ButtonClick" Click="Button_OnClick" />
      <TextBlock x:Name="HtmlContent" Text="{Binding HtmlContent, Mode=TwoWay}"/>
  </StackPanel>
```
- MainView.axaml.cs
```
public partial class MainView : UserControl
{
    public MainView()
    {
        InitializeComponent();
    }

    private async void Button_OnClick(object? sender, RoutedEventArgs e)
    {
        var response = await new HttpClient().GetAsync("google.com");
        if (response.IsSuccessStatusCode)
        {
            HtmlContent.Text = await response.Content.ReadAsStringAsync();
        }
    }
}
```
- MainViewModel
```
public partial class MainViewModel : ObservableObject
{
    [ObservableProperty]
    private string _htmlContent;
    
    public async Task CallAsync()
    {
        var response = await new HttpClient().GetAsync("google.com");
        if (response.IsSuccessStatusCode)
        {
            HtmlContent = await response.Content.ReadAsStringAsync();
        }
    }
}
```

**Comment by maxkatz6** [comment]
Do you have "runMain" or "runMainAndExit" in the `main.js` file?
In the newest templates it was changed "runMain" due to breaking change in .NET 8.
"runMainAndExit" will stop your app right after initialization, which might break async.

**Reply by rincew1nd** [reply]
It is the latest template that was been available a week or two ago. Content of main.js:
```
import { dotnet } from './_framework/dotnet.js'

const is_browser = typeof window != "undefined";
if (!is_browser) throw new Error(`Expected to be running in a browser`);

const dotnetRuntime = await dotnet
    .withDiagnosticTracing(false)
    .withApplicationArgumentsFromQuery()
    .create();

const config = dotnetRuntime.getConfig();

await dotnetRuntime.runMain(config.mainAssemblyName, [window.location.search]);
```

**Reply by rincew1nd** [reply]
I decided to upload whole test project to git repo, here is it - https://github.com/rincew1nd/avalonia-test-app

**Reply by maxkatz6** [reply]
It fails with "invalid request"
![image](https://github.com/user-attachments/assets/d9619986-bd1d-493d-a8a7-3f63d9cf5681)

**Reply by maxkatz6** [reply, marked as answer]
If I change it to `        var response = await new HttpClient().GetAsync("https://www.google.com");` now it "works by design".
I.e. shows another error which is expected:
```
Access to fetch at 'https://www.google.com/' from origin 'https://localhost:5002' has been blocked by CORS policy: No 'Access-Control-Allow-Origin' header is present on the requested resource. If an opaque response serves your needs, set the request's mode to 'no-cors' to fetch the resource with CORS disabled.
```

You can't just request another website or webserver from your browser app. You need to configure CORS on your server to explicitly allow your client origin/domain.

**Reply by rincew1nd** [reply]
Oh, now i get it. We are now talking not a backend application, but a frontend one with it's requirements. I haven't thought about it.
Thank you! Really appreciate a help.
Also, how did you managed to get unhandled exception visible in console? Mine is just empty without any kind of logs.
That's probably why I asked a question in the first place, cause I wouldn't if i saw those errors.

UPD: `.LogToTrace()` made logs to appear, but only for a first hit. So I guess the topic is solved.
